ksqldb-engine/src/main/java/io/confluent/ksql/function/udaf/offset/EarliestByOffset.java

```diff
@@ -123,4 +155,54 @@ public T map(final Struct agg) {
       }
     };
   }
+
+
+  @UdafFactory(description = "Earliest N by offset")
+  static <T> Udaf<T, List<Struct>, List<T>> earliestN(final Schema structSchema,
+      final int topNSize) {
+    return new Udaf<T, List<Struct>, List<T>>() {
+      @Override
+      public List<Struct> initialize() {
+        return new ArrayList<Struct>();
+      }
+
+      @Override
+      public List<Struct> aggregate(final T current, final List<Struct> aggregate) {
+        final int currentSize = aggregate.size();
+        if (currentSize == topNSize) {
+          return aggregate;
+        } else {
+
+          if (current == null) {
+            return aggregate;
+          }
+
+          aggregate.add(createStruct(structSchema, current));
+          return aggregate.subList(0, aggregate.size());
+        }
+
+
+      }
+
+      @Override
+      public List<Struct> merge(final List<Struct> aggOne, final List<Struct> aggTwo) {
+        final List<Struct> merged = new ArrayList<>();
+        merged.addAll(aggOne);
+        merged.addAll(aggTwo);
+        final int currentSize = merged.size();
+        Collections.sort(merged, structComparator);
+        if (currentSize <= topNSize) {
+          return merged;
+        } else {
+          return merged.subList(0, topNSize);
+        }
+      }
```

Comment:
Can pre-allocate the array and simplify the code:

```suggestion
      @Override
      public List<Struct> merge(final List<Struct> aggOne, final List<Struct> aggTwo) {
        final List<Struct> merged = new ArrayList<>(aggOne.size() + aggTwo.size());
        merged.addAll(aggOne);
        merged.addAll(aggTwo);
        merged.sort(structComparator);
        return merged.subList(0, Math.min(topNSize, merged.size()));
      }
```

Plus, I noticed there's not test to cover the code path where the merged array is smaller than `topNSize`. Worth adding that.

Reply:
added